File: src/main/scala/otserver4s/Packet.scala
package otserver4s

import java.io.InputStream

case class TipoRequest(codigo: Byte, descricao: String)
object TipoRequest {
  val LOGIN_REQUEST = TipoRequest(0x01, "LoginRequest")
  val PROCESSAR_LOGIN = TipoRequest(0x0a, "ProcessarLogin")
  val values = Array(LOGIN_REQUEST, PROCESSAR_LOGIN)
  val INVALIDO = TipoRequest(-1, "Inválido")
  def tipoPorCodigo(codigo: Integer) =
    values.filter(_.codigo == codigo).headOption match {
    case Some(ok) => ok
    case None => INVALIDO
  }
}

case class Packet(
    private var posicao: Int = 2, 
    private var buffer: Array[Byte] = Array.ofDim[Byte](0xffff)) {
  def escreverByte(dado: Byte) = {
    buffer(posicao) = dado
    posicao += 1
  }
  def escreverInt16(num: Short) = {
    escreverByte(( num & 0x00ff      ).toByte)
    escreverByte(((num & 0xff00) >> 8).toByte)
  }
  def escreverInt32(num: Long) = {
    escreverByte( (num & 0x000000ff       ).toByte)
    escreverByte(((num & 0x0000ff00) >> 8 ).toByte)
    escreverByte(((num & 0x00ff0000) >> 16).toByte)
    escreverByte(((num & 0xff000000) >> 24).toByte)
  }
  def escreverBytes(dados: Array[Byte]) =
    for(i <- 0 until dados.length) buffer(posicao) = dados(i)
  def escreverString(str: String) = {
    escreverInt16(str.length.toShort)
    str.toCharArray.foreach(c => escreverByte(c.toByte))
  }
  def enviar(socket: java.net.Socket) = {
    val tamanho = (posicao - 2).toShort
    posicao = 0
    escreverInt16(tamanho)
    val pw = new java.io.PrintWriter(socket.getOutputStream, true)
    pw.write(buffer.splitAt(tamanho)._1.map(_.toChar))
    pw.close
  }
}
object Packet {
  val MARCACAO_PACKET_INICIO_LISTA_PERSONAGENS: Byte = 0x64
  val CODIGO_PACKET_LOGIN_SUCESSO: Byte = 0x14
  val CODIGO_PACKET_LOGIN_ERRO: Byte = 0x0a
  val CODIGO_PACKET_PROCESSAR_LOGIN_SUCESSO: Byte = 0x0a
  val CODIGO_PACKET_PROCESSAR_LOGIN_ERRO: Byte = 0x14
  def lerByte(input: InputStream) = input.read
  def lerInt16(input: InputStream) = lerByte(input) | lerByte(input) << 8
  def lerInt32(input: InputStream) = lerByte(input)       | lerByte(input) << 8 |
                                     lerByte(input) << 16 | lerByte(input) << 24
  def lerString(input: InputStream) = {
    val array = Array.ofDim[Byte](Packet.lerInt16(input))
    Option(array).filter(_.length > 0)
      .map(array => array.map(b => Packet.lerByte(input).toByte))
      .map(new String(_))
  }
  def pularLeitura(input: InputStream, n: Long) = input.skip(n)
  private def criarPacketGenericoMensagem(codigo: Byte, mensagem: String) = {
    val packet = Packet()
    packet.escreverByte(codigo)
    packet.escreverString(mensagem)
    packet
  }
  def criarPacketErroLogin(mensagemErro: String) = 
    Packet.criarPacketGenericoMensagem(
      Packet.CODIGO_PACKET_LOGIN_ERRO, mensagemErro)
  def criarPacketProcessarLoginErro(mensagemErro: String) = 
    Packet.criarPacketGenericoMensagem(
      Packet.CODIGO_PACKET_PROCESSAR_LOGIN_ERRO, mensagemErro)
}

File: src/main/kotlin/otkserver/Packet.kt
package otkserver

import java.io.ByteArrayOutputStream
import java.io.InputStream
import org.apache.mina.core.buffer.IoBuffer
import org.apache.mina.core.session.IoSession
import org.apache.mina.core.write.WriteRequest

data class Packet(
  var tamanho: Int = 0,
  val buffer: ByteArray = ByteArray(0xffff - 2) { 0x00 }
) {
  companion object {
    val MARCACAO_PACKET_INICIO_LISTA_PERSONAGENS = 0x64
    val CODIGO_PACKET_LOGIN_SUCESSO = 0x14
    val CODIGO_PACKET_LOGIN_ERRO = 0x0a
    val CODIGO_PACKET_PROCESSAR_LOGIN_SUCESSO = 0x0a
    val CODIGO_PACKET_PROCESSAR_LOGIN_ERRO = 0x14
    fun lerByte(input: InputStream)  = input.read() and 0xff
    fun lerInt16(input: InputStream) = lerByte(input) or (lerByte(input) shl 8)
    fun lerInt32(input: InputStream) = lerByte(input) or
                                    (lerByte(input) shl 8) or
                                    (lerByte(input) shl 16) or
                                    (lerByte(input) shl 24)
    fun lerString(input: InputStream): String? {
      val saida =
		    String(ByteArray(Packet.lerInt16(input)) { Packet.lerByte(input).toByte() })
		  return if(saida.length > 0) saida else null
	  }
    fun pularLeitura(input: InputStream, n: Long) = input.skip(n)
	  private fun criarPacketGenericoMensagem(
		    codigo: Int, mensagem: String): Packet {
      val packet = Packet()
      packet.escreverByte(codigo)
      packet.escreverString(mensagem)
      return packet
    }
    fun criarPacketErroLogin(mensagemErro: String) = 
      Packet.criarPacketGenericoMensagem(
        Packet.CODIGO_PACKET_LOGIN_ERRO, mensagemErro)
    fun criarPacketProcessarLoginErro(mensagemErro: String) = 
      Packet.criarPacketGenericoMensagem(
        Packet.CODIGO_PACKET_PROCESSAR_LOGIN_ERRO, mensagemErro)
  }
	fun escreverByte(_byte: Byte) {
	  buffer[tamanho++] = _byte
  }
  fun escreverByte(_byte: Int) {
	  escreverByte((_byte and 0xff).toByte())
  }
	fun escreverByte(_byte: Char) {
		escreverByte(_byte.toByte())
	}
  fun escreverInt16(_int: Int) {
    escreverByte(_int and 0x00ff)
    escreverByte((_int and 0xff00) shr 8)
  }
  fun escreverInt32(_long: Long) {
    escreverByte( (_long and 0x000000ff).toInt())
    escreverByte(((_long and 0x0000ff00) shr 8).toInt())
    escreverByte(((_long and 0x00ff0000) shr 16).toInt())
    escreverByte(((_long and 0xff000000) shr 24).toInt())
  }
  fun escreverString(_str: String) {
    escreverInt16(_str.length)
    _str.toCharArray().forEach { escreverByte(it) }
  }
	fun getBytesComTamanho(): ByteArray {
		return byteArrayOf((tamanho and 0x00ff).toByte(),
		  ((tamanho and 0xff00) shr 8).toByte()) + buffer
	}
  fun enviar(sessao: IoSession, desconectar: Boolean) {
	  val ioBuffer = IoBuffer.allocate(tamanho, false)
	  ioBuffer.setAutoExpand(true)
	  getBytesComTamanho().forEach { ioBuffer.put(it) }
	  ioBuffer.flip()
	  sessao.write(ioBuffer)
	  ioBuffer.free()
	  if(desconectar) sessao.closeOnFlush()
  }
}
